File: lunar_fishing.cpp
// lunar_fishing.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <map>
#include <cmath>
#include <ctime>
#include <iomanip>
#include <sstream>
#include <nlohmann/json.hpp>
#include <getopt.h>

using namespace std;
using json = nlohmann::json;

const double DEFAULT_LAT = 0.0;
const double DEFAULT_LON = 0.0;
const string CONFIG_FILE = "lunar_fishing_config.json";

struct Location {
    double lat, lon;
};

struct MoonData {
    string phase;
    double illumination;
    double age;
    string zodiac;
};

struct SolunarData {
    double major1Start, major1End, major2Start, major2End;
    double minor1Start, minor1End, minor2Start, minor2End;
    double moonTransit;
};

double julianDay(const tm& dt) {
    int year = dt.tm_year + 1900;
    int month = dt.tm_mon + 1;
    double day = dt.tm_mday + dt.tm_hour/24.0 + dt.tm_min/1440.0 + dt.tm_sec/86400.0;
    if (month <= 2) { year--; month += 12; }
    int A = year / 100;
    int B = 2 - A + A / 4;
    return (int)(365.25 * (year + 4716)) + (int)(30.6001 * (month + 1)) + day + B - 1524.5;
}

pair<double, double> moonPosition(double jd) {
    double T = (jd - 2451545.0) / 36525.0;
    double L_prime = 218.3165 + 481267.8813 * T;
    double D = 297.8502 + 445267.1114 * T;
    double M = 357.5291 + 35999.0503 * T;
    double M_prime = 134.9634 + 477198.8676 * T;
    double F = 93.2720 + 483202.0175 * T;

    L_prime = fmod(L_prime, 360) * M_PI / 180;
    D = fmod(D, 360) * M_PI / 180;
    M = fmod(M, 360) * M_PI / 180;
    M_prime = fmod(M_prime, 360) * M_PI / 180;
    F = fmod(F, 360) * M_PI / 180;

    double lon = L_prime + (6.289 * sin(M_prime) + 1.274 * sin(2*D - M_prime) + 0.658 * sin(2*D) + 0.214 * sin(2*M_prime) - 0.186 * sin(M) - 0.114 * sin(2*F)) * M_PI / 180;
    double lat = (5.128 * sin(F) + 0.280 * sin(M_prime + F) + 0.278 * sin(M_prime - F) + 0.173 * sin(2*D - F)) * M_PI / 180;
    return {lon, lat};
}

double sunPosition(double jd) {
    double T = (jd - 2451545.0) / 36525.0;
    double M = fmod(357.5291 + 35999.0503 * T, 360) * M_PI / 180;
    double C = 1.9146 * sin(M) + 0.0200 * sin(2*M) + 0.0003 * sin(3*M);
    double lon = fmod(280.4665 + 36000.7698 * T + C, 360) * M_PI / 180;
    return lon;
}

double greenwichSiderealTime(double jd) {
    double T = (jd - 2451545.0) / 36525.0;
    double gmst = 280.46061837 + 360.98564736629 * (jd - 2451545.0) + 0.000387933 * T*T - T*T*T / 38710000.0;
    gmst = fmod(gmst, 360.0);
    if (gmst < 0) gmst += 360.0;
    return gmst / 15.0;
}

MoonData moonPhase(const tm& dt) {
    double jd = julianDay(dt);
    auto [lonMoon, _] = moonPosition(jd);
    double lonSun = sunPosition(jd);

    double elong = lonMoon - lonSun;
    elong = atan2(sin(elong), cos(elong));
    double phaseAngle = atan2(sin(elong), cos(elong));
    double illumination = (1 + cos(phaseAngle)) / 2;

    double age = (jd - 2451550.1) / 29.53058867;
    age = fmod(age, 29.53058867);
    if (age < 0) age += 29.53058867;

    string phase;
    if (age < 1.0) phase = "New Moon";
    else if (age < 7.38) phase = "Waxing Crescent";
    else if (age < 8.38) phase = "First Quarter";
    else if (age < 14.77) phase = "Waxing Gibbous";
    else if (age < 15.77) phase = "Full Moon";
    else if (age < 22.15) phase = "Waning Gibbous";
    else if (age < 23.15) phase = "Last Quarter";
    else phase = "Waning Crescent";

    vector<string> signs = {"Aries","Taurus","Gemini","Cancer","Leo","Virgo","Libra","Scorpio","Sagittarius","Capricorn","Aquarius","Pisces"};
    double lonDeg = fmod(lonMoon * 180 / M_PI, 360);
    if (lonDeg < 0) lonDeg += 360;
    int idx = (int)(lonDeg / 30);
    string zodiac = signs[idx];

    return {phase, illumination * 100, age, zodiac};
}

SolunarData solunar(const tm& dt, double lat, double lon) {
    double jd = julianDay(dt);
    auto [lonMoon, _] = moonPosition(jd);
    double lonMoonDeg = fmod(lonMoon * 180 / M_PI, 360);
    if (lonMoonDeg < 0) lonMoonDeg += 360;
    double gst = greenwichSiderealTime(jd);
    double lst = fmod(gst + lon / 15.0, 24.0);
    if (lst < 0) lst += 24.0;
    double moonRA = lonMoonDeg / 15.0;
    double moonTransit = fmod(moonRA - lst, 24.0);
    if (moonTransit < 0) moonTransit += 24.0;

    return {
        moonTransit - 1.0, moonTransit + 1.0,
        moonTransit + 12.0 - 1.0, moonTransit + 12.0 + 1.0,
        moonTransit + 6.0 - 0.5, moonTransit + 6.0 + 0.5,
        moonTransit + 18.0 - 0.5, moonTransit + 18.0 + 0.5,
        moonTransit
    };
}

int fishingRating(const string& phase) {
    if (phase == "Full Moon" || phase == "New Moon") return 4;
    if (phase == "Waxing Gibbous" || phase == "Waning Gibbous" || phase == "First Quarter" || phase == "Last Quarter") return 3;
    if (phase == "Waxing Crescent" || phase == "Waning Crescent") return 2;
    return 1;
}

string formatTime(double hours) {
    int h = (int)hours;
    int m = (int)((hours - h) * 60);
    char buf[6];
    snprintf(buf, sizeof(buf), "%02d:%02d", h, m);
    return string(buf);
}

void render(const tm& dt, double lat, double lon, bool phaseOnly) {
    MoonData moonData = moonPhase(dt);
    SolunarData solunarData = solunar(dt, lat, lon);
    int rating = fishingRating(moonData.phase);

    if (phaseOnly) {
        cout << moonData.phase << "\n";
        return;
    }

    vector<string> ratings = {"Poor", "Fair", "Good", "Excellent"};
    vector<string> ratingEmoji = {"⭐", "⭐⭐", "⭐⭐⭐", "⭐⭐⭐⭐"};

    cout << "\n🎣 Lunar Fishing Calendar\n";
    string latStr = to_string(abs(lat)).substr(0,5) + "°" + (lat >= 0 ? "N" : "S");
    string lonStr = to_string(abs(lon)).substr(0,5) + "°" + (lon >= 0 ? "E" : "W");
    cout << "Location: " << latStr << ", " << lonStr << "\n";
    char dateBuf[20];
    strftime(dateBuf, sizeof(dateBuf), "%Y-%m-%d %H:%M", &dt);
    cout << "Date: " << dateBuf << "\n";

    cout << "\n🌓 Moon Phase: " << moonData.phase << " (" << fixed << setprecision(1) << moonData.illumination << "% illuminated)\n";
    cout << "🌙 Moon Age: " << fixed << setprecision(1) << moonData.age << " days\n";
    cout << "♑ Zodiac: " << moonData.zodiac << "\n";

    cout << "\n🎯 Solunar Feeding Periods:\n";
    struct Period { double start, end; string label; string emoji; string rating; };
    vector<Period> periods = {
        {solunarData.major1Start, solunarData.major1End, "Major Period 1", ratingEmoji[rating-1], ratings[rating-1]},
        {solunarData.major2Start, solunarData.major2End, "Major Period 2", ratingEmoji[rating-1], ratings[rating-1]},
        {solunarData.minor1Start, solunarData.minor1End, "Minor Period 1", "⭐", "Good"},
        {solunarData.minor2Start, solunarData.minor2End, "Minor Period 2", "⭐", "Good"},
    };
    for (auto& p : periods) {
        if (p.start >= 0 && p.start < 24) {
            cout << "  " << p.label << ": " << formatTime(p.start) << " – " << formatTime(p.end) << " (" << p.emoji << " " << p.rating << ")\n";
        }
    }

    cout << "\n⭐ Best Fishing Rating: " << ratings[rating-1] << " (" << ratingEmoji[rating-1] << ")\n";

    // Sunrise/sunset (simplified)
    int dayOfYear = dt.tm_yday + 1;
    double declination = 23.44 * sin((284 + dayOfYear) * 360 * M_PI / 180 / 365);
    double latRad = lat * M_PI / 180;
    double decRad = declination * M_PI / 180;
    double cosHA = -tan(latRad) * tan(decRad);
    double ha;
    if (cosHA < -1) ha = M_PI;
    else if (cosHA > 1) ha = 0;
    else ha = acos(cosHA);
    double dayLength = ha * 2 / (M_PI / 12);
    double noon = 12.0 - lon / 15.0;
    double sunrise = noon - dayLength / 2;
    double sunset = noon + dayLength / 2;
    cout << "🌅 Sunrise: " << formatTime(sunrise) << " | 🌇 Sunset: " << formatTime(sunset) << "\n";

    double moonRise = fmod(solunarData.moonTransit + 6, 24.0);
    double moonSet = fmod(solunarData.moonTransit + 18, 24.0);
    cout << "🌙 Moonrise: " << formatTime(moonRise) << " | 🌇 Moonset: " << formatTime(moonSet) << "\n";
}

int main(int argc, char* argv[]) {
    static struct option long_options[] = {
        {"date", required_argument, 0, 'd'},
        {"lat", required_argument, 0, 'a'},
        {"lon", required_argument, 0, 'o'},
        {"phase-only", no_argument, 0, 'p'},
        {"save-location", required_argument, 0, 's'},
        {"use-location", required_argument, 0, 'u'},
        {"list-locations", no_argument, 0, 'l'},
        {0,0,0,0}
    };
    int opt;
    string dateStr, saveLocation, useLocation;
    double lat = DEFAULT_LAT, lon = DEFAULT_LON;
    bool phaseOnly = false, listLocations = false;

    while ((opt = getopt_long(argc, argv, "d:a:o:ps:u:l", long_options, nullptr)) != -1) {
        switch (opt) {
            case 'd': dateStr = optarg; break;
            case 'a': lat = stod(optarg); break;
            case 'o': lon = stod(optarg); break;
            case 'p': phaseOnly = true; break;
            case 's': saveLocation = optarg; break;
            case 'u': useLocation = optarg; break;
            case 'l': listLocations = true; break;
            default:
                cerr << "Usage: lunar_fishing --date YYYY-MM-DD --lat LAT --lon LON --phase-only --save-location NAME --use-location NAME --list-locations\n";
                return 1;
        }
    }

    // Load config
    json config;
    ifstream f(CONFIG_FILE);
    if (f.is_open()) {
        f >> config;
    }

    if (listLocations) {
        if (!config.contains("locations") || config["locations"].empty()) {
            cout << "No saved locations.\n";
            return 0;
        }
        cout << "\n📍 Saved Locations:\n";
        for (auto& [name, loc] : config["locations"].items()) {
            string latStr = to_string(abs(loc["lat"].get<double>())).substr(0,5) + "°" + (loc["lat"].get<double>() >= 0 ? "N" : "S");
            string lonStr = to_string(abs(loc["lon"].get<double>())).substr(0,5) + "°" + (loc["lon"].get<double>() >= 0 ? "E" : "W");
            cout << "  " << name << ": " << latStr << ", " << lonStr << "\n";
        }
        return 0;
    }

    if (!saveLocation.empty()) {
        config["locations"][saveLocation] = {{"lat", lat}, {"lon", lon}};
        ofstream out(CONFIG_FILE);
        out << setw(2) << config << endl;
        cout << "✅ Location '" << saveLocation << "' saved.\n";
    }

    if (!useLocation.empty()) {
        if (config.contains("locations") && config["locations"].contains(useLocation)) {
            lat = config["locations"][useLocation]["lat"];
            lon = config["locations"][useLocation]["lon"];
            cout << "📍 Using saved location: " << useLocation << "\n";
        } else {
            cout << "Location '" << useLocation << "' not found.\n";
            return 1;
        }
    }

    time_t now = time(nullptr);
    tm dt = *gmtime(&now);
    if (!dateStr.empty()) {
        strptime(dateStr.c_str(), "%Y-%m-%d", &dt);
    }

    render(dt, lat, lon, phaseOnly);
    return 0;
}

File: lunar_fishing.js
// lunar_fishing.js
#!/usr/bin/env node
const fs = require('fs');
const { program } = require('commander');

const DEFAULT_LAT = 0.0;
const DEFAULT_LON = 0.0;
const CONFIG_FILE = 'lunar_fishing_config.json';

function julianDay(date) {
    let year = date.getFullYear();
    let month = date.getMonth() + 1;
    let day = date.getDate() + date.getHours()/24 + date.getMinutes()/1440 + date.getSeconds()/86400;
    if (month <= 2) { year--; month += 12; }
    let A = Math.floor(year / 100);
    let B = 2 - A + Math.floor(A / 4);
    return Math.floor(365.25 * (year + 4716)) + Math.floor(30.6001 * (month + 1)) + day + B - 1524.5;
}

function moonPosition(jd) {
    let T = (jd - 2451545.0) / 36525.0;
    let L_prime = 218.3165 + 481267.8813 * T;
    let D = 297.8502 + 445267.1114 * T;
    let M = 357.5291 + 35999.0503 * T;
    let M_prime = 134.9634 + 477198.8676 * T;
    let F = 93.2720 + 483202.0175 * T;

    L_prime = ((L_prime % 360) + 360) % 360 * Math.PI / 180;
    D = ((D % 360) + 360) % 360 * Math.PI / 180;
    M = ((M % 360) + 360) % 360 * Math.PI / 180;
    M_prime = ((M_prime % 360) + 360) % 360 * Math.PI / 180;
    F = ((F % 360) + 360) % 360 * Math.PI / 180;

    let lon = L_prime + (6.289 * Math.sin(M_prime) + 1.274 * Math.sin(2*D - M_prime) + 0.658 * Math.sin(2*D) + 0.214 * Math.sin(2*M_prime) - 0.186 * Math.sin(M) - 0.114 * Math.sin(2*F)) * Math.PI / 180;
    let lat = (5.128 * Math.sin(F) + 0.280 * Math.sin(M_prime + F) + 0.278 * Math.sin(M_prime - F) + 0.173 * Math.sin(2*D - F)) * Math.PI / 180;
    return { lon, lat };
}

function sunPosition(jd) {
    let T = (jd - 2451545.0) / 36525.0;
    let M = ((357.5291 + 35999.0503 * T) % 360) * Math.PI / 180;
    let C = 1.9146 * Math.sin(M) + 0.0200 * Math.sin(2*M) + 0.0003 * Math.sin(3*M);
    let lon = ((280.4665 + 36000.7698 * T + C) % 360) * Math.PI / 180;
    return lon;
}

function greenwichSiderealTime(jd) {
    let T = (jd - 2451545.0) / 36525.0;
    let gmst = 280.46061837 + 360.98564736629 * (jd - 2451545.0) + 0.000387933 * T*T - T*T*T / 38710000.0;
    gmst = ((gmst % 360) + 360) % 360;
    return gmst / 15.0;
}

function moonPhase(date) {
    let jd = julianDay(date);
    let { lon: lonMoon } = moonPosition(jd);
    let lonSun = sunPosition(jd);

    let elong = lonMoon - lonSun;
    elong = Math.atan2(Math.sin(elong), Math.cos(elong));
    let phaseAngle = Math.atan2(Math.sin(elong), Math.cos(elong));
    let illumination = (1 + Math.cos(phaseAngle)) / 2;

    let age = (jd - 2451550.1) / 29.53058867;
    age = ((age % 29.53058867) + 29.53058867) % 29.53058867;

    let phase;
    if (age < 1.0) phase = "New Moon";
    else if (age < 7.38) phase = "Waxing Crescent";
    else if (age < 8.38) phase = "First Quarter";
    else if (age < 14.77) phase = "Waxing Gibbous";
    else if (age < 15.77) phase = "Full Moon";
    else if (age < 22.15) phase = "Waning Gibbous";
    else if (age < 23.15) phase = "Last Quarter";
    else phase = "Waning Crescent";

    const signs = ["Aries","Taurus","Gemini","Cancer","Leo","Virgo","Libra","Scorpio","Sagittarius","Capricorn","Aquarius","Pisces"];
    let lonDeg = ((lonMoon * 180 / Math.PI) % 360 + 360) % 360;
    let idx = Math.floor(lonDeg / 30);
    let zodiac = signs[idx];

    return { phase, illumination: illumination * 100, age, zodiac, lonMoon, jd };
}

function solunar(date, lat, lon) {
    let jd = julianDay(date);
    let { lon: lonMoon } = moonPosition(jd);
    let lonMoonDeg = ((lonMoon * 180 / Math.PI) % 360 + 360) % 360;
    let gst = greenwichSiderealTime(jd);
    let lst = ((gst + lon / 15.0) % 24.0 + 24.0) % 24.0;
    let moonRA = lonMoonDeg / 15.0;
    let moonTransit = ((moonRA - lst) % 24.0 + 24.0) % 24.0;
    let major1Start = moonTransit - 1.0;
    let major1End = moonTransit + 1.0;
    let major2Start = moonTransit + 12.0 - 1.0;
    let major2End = moonTransit + 12.0 + 1.0;
    let minor1Start = moonTransit + 6.0 - 0.5;
    let minor1End = moonTransit + 6.0 + 0.5;
    let minor2Start = moonTransit + 18.0 - 0.5;
    let minor2End = moonTransit + 18.0 + 0.5;
    return { major1Start, major1End, major2Start, major2End, minor1Start, minor1End, minor2Start, minor2End, moonTransit };
}

function fishingRating(phase) {
    switch (phase) {
        case "Full Moon": case "New Moon": return 4;
        case "Waxing Gibbous": case "Waning Gibbous": case "First Quarter": case "Last Quarter": return 3;
        case "Waxing Crescent": case "Waning Crescent": return 2;
        default: return 1;
    }
}

function formatTime(hours) {
    const h = Math.floor(hours);
    const m = Math.floor((hours - h) * 60);
    return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}`;
}

function render(date, lat, lon, phaseOnly) {
    let moonData = moonPhase(date);
    let solunarData = solunar(date, lat, lon);
    let rating = fishingRating(moonData.phase);

    if (phaseOnly) {
        console.log(moonData.phase);
        return;
    }

    const ratings = ["Poor", "Fair", "Good", "Excellent"];
    const ratingEmoji = ["⭐", "⭐⭐", "⭐⭐⭐", "⭐⭐⭐⭐"];

    console.log(`\n🎣 Lunar Fishing Calendar`);
    const latStr = `${Math.abs(lat).toFixed(2)}°${lat >= 0 ? 'N' : 'S'}`;
    const lonStr = `${Math.abs(lon).toFixed(2)}°${lon >= 0 ? 'E' : 'W'}`;
    console.log(`Location: ${latStr}, ${lonStr}`);
    console.log(`Date: ${date.toISOString().slice(0,16).replace('T',' ')}`);

    console.log(`\n🌓 Moon Phase: ${moonData.phase} (${moonData.illumination.toFixed(1)}% illuminated)`);
    console.log(`🌙 Moon Age: ${moonData.age.toFixed(1)} days`);
    console.log(`♑ Zodiac: ${moonData.zodiac}`);

    console.log(`\n🎯 Solunar Feeding Periods:`);
    const periods = [
        { start: solunarData.major1Start, end: solunarData.major1End, label: 'Major Period 1', emoji: ratingEmoji[rating-1], rating: ratings[rating-1] },
        { start: solunarData.major2Start, end: solunarData.major2End, label: 'Major Period 2', emoji: ratingEmoji[rating-1], rating: ratings[rating-1] },
        { start: solunarData.minor1Start, end: solunarData.minor1End, label: 'Minor Period 1', emoji: '⭐', rating: 'Good' },
        { start: solunarData.minor2Start, end: solunarData.minor2End, label: 'Minor Period 2', emoji: '⭐', rating: 'Good' },
    ];
    for (const p of periods) {
        if (p.start >= 0 && p.start < 24) {
            console.log(`  ${p.label}: ${formatTime(p.start)} – ${formatTime(p.end)} (${p.emoji} ${p.rating})`);
        }
    }

    console.log(`\n⭐ Best Fishing Rating: ${ratings[rating-1]} (${ratingEmoji[rating-1]})`);

    // Sunrise/sunset (simplified)
    const dayOfYear = Math.floor((date - new Date(date.getFullYear(), 0, 0)) / (1000*60*60*24));
    const declination = 23.44 * Math.sin((284 + dayOfYear) * 360 * Math.PI / 180 / 365);
    const latRad = lat * Math.PI / 180;
    const decRad = declination * Math.PI / 180;
    let cosHA = -Math.tan(latRad) * Math.tan(decRad);
    let ha;
    if (cosHA < -1) ha = Math.PI;
    else if (cosHA > 1) ha = 0;
    else ha = Math.acos(cosHA);
    const dayLength = ha * 2 / (Math.PI / 12);
    const noon = 12.0 - lon / 15.0;
    const sunrise = noon - dayLength / 2;
    const sunset = noon + dayLength / 2;
    console.log(`🌅 Sunrise: ${formatTime(sunrise)} | 🌇 Sunset: ${formatTime(sunset)}`);

    const moonRise = (solunarData.moonTransit + 6) % 24;
    const moonSet = (solunarData.moonTransit + 18) % 24;
    console.log(`🌙 Moonrise: ${formatTime(moonRise)} | 🌇 Moonset: ${formatTime(moonSet)}`);
}

program
    .option('--date <date>', 'YYYY-MM-DD')
    .option('--lat <lat>', 'Latitude (positive North)', parseFloat, DEFAULT_LAT)
    .option('--lon <lon>', 'Longitude (positive East)', parseFloat, DEFAULT_LON)
    .option('--phase-only', 'Output only phase name')
    .option('--save-location <name>', 'Save location with name')
    .option('--use-location <name>', 'Use saved location')
    .option('--list-locations', 'List saved locations')
    .parse(process.argv);

const opts = program.opts();

// Load config
let config = { locations: {} };
if (fs.existsSync(CONFIG_FILE)) {
    try {
        config = JSON.parse(fs.readFileSync(CONFIG_FILE));
    } catch (e) {}
}

if (opts.listLocations) {
    if (Object.keys(config.locations).length === 0) {
        console.log('No saved locations.');
        process.exit(0);
    }
    console.log('\n📍 Saved Locations:');
    for (const [name, loc] of Object.entries(config.locations)) {
        const latStr = `${Math.abs(loc.lat).toFixed(2)}°${loc.lat >= 0 ? 'N' : 'S'}`;
        const lonStr = `${Math.abs(loc.lon).toFixed(2)}°${loc.lon >= 0 ? 'E' : 'W'}`;
        console.log(`  ${name}: ${latStr}, ${lonStr}`);
    }
    process.exit(0);
}

if (opts.saveLocation) {
    config.locations[opts.saveLocation] = { lat: opts.lat, lon: opts.lon };
    fs.writeFileSync(CONFIG_FILE, JSON.stringify(config, null, 2));
    console.log(`✅ Location '${opts.saveLocation}' saved.`);
}

let lat = opts.lat || DEFAULT_LAT;
let lon = opts.lon || DEFAULT_LON;

if (opts.useLocation) {
    if (config.locations[opts.useLocation]) {
        lat = config.locations[opts.useLocation].lat;
        lon = config.locations[opts.useLocation].lon;
        console.log(`📍 Using saved location: ${opts.useLocation}`);
    } else {
        console.log(`Location '${opts.useLocation}' not found.`);
        process.exit(1);
    }
}

let dt = new Date();
if (opts.date) {
    dt = new Date(opts.date + 'T00:00:00');
}
render(dt, lat, lon, opts.phaseOnly);
